contrib/golang/filters/http/source/go/pkg/http/shim.go

```diff
@@ -276,14 +282,67 @@ func envoyGoFilterOnHttpLog(r *C.httpRequest, logType uint64) {
 
 	v := api.AccessLogType(logType)
 
+	// Request headers must exist because the HTTP filter won't be run if the headers are
+	// not sent yet.
+	// TODO: make the headers/trailers read-only
+	reqHeader := &requestHeaderMapImpl{
+		requestOrResponseHeaderMapImpl{
+			headerMapImpl{
+				state:       decodingState,
+				headerNum:   reqHeaderNum,
+				headerBytes: reqHeaderBytes,
+			},
+		},
+	}
+
+	var reqTrailer api.RequestTrailerMap
+	if reqTrailerNum != 0 {
+		reqTrailer = &requestTrailerMapImpl{
+			requestOrResponseTrailerMapImpl{
+				headerMapImpl{
+					state:       decodingState,
+					headerNum:   reqTrailerNum,
+					headerBytes: reqTrailerBytes,
+				},
+			},
+		}
+	}
+
+	var respHeader api.ResponseHeaderMap
+	if respHeaderNum != 0 {
+		respHeader = &responseHeaderMapImpl{
+			requestOrResponseHeaderMapImpl{
+				headerMapImpl{
+					state:       encodingState,
+					headerNum:   respHeaderNum,
+					headerBytes: respHeaderBytes,
+				},
+			},
+		}
+	}
+
+	var respTrailer api.ResponseTrailerMap
+	if respTrailerNum != 0 {
+		respTrailer = &responseTrailerMapImpl{
+			requestOrResponseTrailerMapImpl{
+				headerMapImpl{
+					state:       encodingState,
+					headerNum:   respTrailerNum,
+					headerBytes: respTrailerBytes,
+				},
+			},
+		}
+	}
+
 	f := req.httpFilter
+
 	switch v {
-	case api.AccessLogDownstreamStart:
-		f.OnLogDownstreamStart()
-	case api.AccessLogDownstreamPeriodic:
-		f.OnLogDownstreamPeriodic()
 	case api.AccessLogDownstreamEnd:
-		f.OnLog()
+		f.OnLog(reqHeader, reqTrailer, respHeader, respTrailer)
```

Comment:
How about use the Envoy C++ API style: 
```
OnLog(logContext)
OnLogDownstreamPeriodic(logContext)
OnLogDownstreamStart(logContext)

type logContext interface {
    requestHeaders() RequestHeaderMap
    ...
}
```
benifits:
1. we do not need to pre create four interface values even people won't use it. (we do only need one instead)
2. API fexible, we can add more APIs into logContext in the feature.

Reply:
@doujiang24 
I wonder how to connect this with the goal in https://github.com/envoyproxy/envoy/issues/33873.
If we add the methods in the logContext, then we will have two different implementations to get request headers:
one can only use in OnLog phase and the other is used in the Encode phase.

Reply:
This is just API style suggestion, no relation to https://github.com/envoyproxy/envoy/issues/33873.
The implementations could keep the same as now, just user face API tweaks.

Reply:
@doujiang24 
It's not just user face API tweaks. It also affects how we implement it internally. If we bind the requestHeaders to the logContext, then we need to get the header from the logContext, which is unavailable in Encode phases. However, if we decide to support RequestHeaders in StreamInfo, and now we have two FilterCallbacks, we will probably fetch the headers from the XXFilterCallback directly.

Reply:
This is my thought:
1. save the all of the parameters into `logContext`, i.e. `decodingState`, `encodingState` ...
2. create a `RequestHeaderMap` in `logContext.requestHeaders`, and etc.

we do not need any C/cgo related changes, just Go code tweaks, thoughts?